# Supervised Learning Model Evaluation Lab

Complete the exercises below to solidify your knowledge and understanding of supervised learning model evaluation.

In [4]:
import pandas as pd

import warnings
# --- Import the 'warnings' library ---
# This is a built-in Python module used to control warning messages.
# Warnings are informational messages (not errors) that alert you about 
# deprecated features or potential issues in your code.

warnings.filterwarnings('ignore')
# --- Filter all warning messages and ignore them ---
# 'filterwarnings' is a function that lets you control which warnings are shown.
# Using the argument 'ignore' tells Python to suppress (hide) all warnings.
# This helps keep your notebook output clean when you're sure the warnings are harmless.

## Regression Model Evaluation

In [6]:
column_names = ['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD', 'TAX', 'PTRATIO', 'B', 'LSTAT', 'MEDV']
data = pd.read_csv('housing.csv', header=None, delimiter=r"\s+", names=column_names)
# Parameters explained:
#   'header=None' → the CSV file doesn’t include a header row, so we’ll use our own column names.
#   'delimiter=r\"\\s+\"' → the 'r' prefix creates a raw string; '\\s+' means “split on one or more spaces”,
#                           since this dataset uses whitespace as the separator instead of commas.
#   'names=column_names' → assigns the list of column names we defined above to the DataFrame.

In [10]:
"""
CRIM - per capita crime rate by town
ZN - proportion of residential land zoned for lots over 25,000 sq.ft.
INDUS - proportion of non-retail business acres per town.
CHAS - Charles River dummy variable (1 if tract bounds river; 0 otherwise)
NOX - nitric oxides concentration (parts per 10 million)
RM - average number of rooms per dwelling
AGE - proportion of owner-occupied units built prior to 1940
DIS - weighted distances to five Boston employment centres
RAD - index of accessibility to radial highways
TAX - full-value property-tax rate per $10,000
PTRATIO - pupil-teacher ratio by town
B - 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town
LSTAT - % lower status of the population
MEDV - Median value of owner-occupied homes in $1000's"""

"\nCRIM - per capita crime rate by town\nZN - proportion of residential land zoned for lots over 25,000 sq.ft.\nINDUS - proportion of non-retail business acres per town.\nCHAS - Charles River dummy variable (1 if tract bounds river; 0 otherwise)\nNOX - nitric oxides concentration (parts per 10 million)\nRM - average number of rooms per dwelling\nAGE - proportion of owner-occupied units built prior to 1940\nDIS - weighted distances to five Boston employment centres\nRAD - index of accessibility to radial highways\nTAX - full-value property-tax rate per $10,000\nPTRATIO - pupil-teacher ratio by town\nB - 1000(Bk - 0.63)^2 where Bk is the proportion of blacks by town\nLSTAT - % lower status of the population\nMEDV - Median value of owner-occupied homes in $1000's"

In [11]:
data

,CRIM,ZN,INDUS,CHAS,NOX,RM,AGE,DIS,RAD,TAX,PTRATIO,B,LSTAT,MEDV
0,0.00632,18.0,2.31,0,0.538,6.575,65.2,4.0900,1,296.0,15.3,396.90,4.98,24.0
1,0.02731,0.0,7.07,0,0.469,6.421,78.9,4.9671,2,242.0,17.8,396.90,9.14,21.6
2,0.02729,0.0,7.07,0,0.469,7.185,61.1,4.9671,2,242.0,17.8,392.83,4.03,34.7
3,0.03237,0.0,2.18,0,0.458,6.998,45.8,6.0622,3,222.0,18.7,394.63,2.94,33.4
4,0.06905,0.0,2.18,0,0.458,7.147,54.2,6.0622,3,222.0,18.7,396.90,5.33,36.2
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
501,0.06263,0.0,11.93,0,0.573,6.593,69.1,2.4786,1,273.0,21.0,391.99,9.67,22.4
502,0.04527,0.0,11.93,0,0.573,6.120,76.7,2.2875,1,273.0,21.0,396.90,9.08,20.6
503,0.06076,0.0,11.93,0,0.573,6.976,91.0,2.1675,1,273.0,21.0,396.90,5.64,23.9
504,0.10959,0.0,11.93,0,0.573,6.794,89.3,2.3889,1,273.0,21.0,393.45,6.48,22.0


## 1. Split this data set into training (80%) and testing (20%) sets.

The `MEDV` field represents the median value of owner-occupied homes (in $1000's) and is the target variable that we will want to predict.

In [12]:
# Your code here :

# --- Import the function to split the dataset ---
# 'train_test_split' is a utility from scikit-learn used to divide data into training and testing sets.
from sklearn.model_selection import train_test_split

# --- Separate the features (X) from the target variable (y) ---
# 'X' will contain all columns except 'MEDV', which are the input features.
# 'y' will contain only 'MEDV', which is what we want to predict.
X = data.drop('MEDV', axis=1)  # axis=1 means “drop column” (not row)
y = data['MEDV']               # target column

# --- Split the data into training (80%) and testing (20%) subsets ---
# Parameters:
#   test_size=0.2 → 20% of the data will be used for testing.
#   random_state=42 → ensures reproducibility (same random split each time).
#   The function automatically shuffles before splitting.
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

# --- Check the resulting shapes to confirm the split ---
print("Training set shape:", X_train.shape)
print("Testing set shape:", X_test.shape)


Training set shape: (404, 13)
Testing set shape: (102, 13)


## 2. Train a `LinearRegression` model on this data set and generate predictions on both the training and the testing set.

In [15]:
# --- Import the LinearRegression model from scikit-learn ---
# LinearRegression is a simple yet powerful supervised learning algorithm
# that models the relationship between a dependent variable (y)
# and one or more independent variables (X) by fitting a straight line.
from sklearn.linear_model import LinearRegression

# --- Create an instance of the LinearRegression model ---
# This initializes an empty model object that will later be trained (fitted) on the data.
# Parameters:
#   fit_intercept=True → the model will calculate the bias term (b₀).
#   n_jobs=None → uses a single CPU core; you can set n_jobs=-1 to use all cores.
lin_reg = LinearRegression()

# --- Train (fit) the model on the training data ---
# .fit() estimates the best-fitting linear equation:
#   y = b₀ + b₁x₁ + b₂x₂ + ... + bₙxₙ
# It does this by minimizing the residual sum of squares (errors between predictions and true values).
lin_reg.fit(X_train, y_train)

# --- Generate predictions on the training set ---
# The model uses the learned coefficients to predict MEDV values for the same data it was trained on.
y_train_pred = lin_reg.predict(X_train)

# --- Generate predictions on the testing set ---
# Here we test how well the model generalizes to unseen data (the 20% test split).
y_test_pred = lin_reg.predict(X_test)

# --- Optional: check the shapes of predictions ---
print("Training predictions shape:", y_train_pred.shape)
print("Testing predictions shape:", y_test_pred.shape)


Training predictions shape: (404,)
Testing predictions shape: (102,)


## 3. Calculate and print R-squared for both the training and the testing set.

In [16]:
# Your code here :

# --- Import the R-squared metric function from scikit-learn ---
# R² (R-squared) measures how well the regression model explains the variability of the target variable.
#   - R² = 1 → perfect prediction (model explains 100% of variance)
#   - R² = 0 → model explains nothing (same as predicting the mean)
#   - R² < 0 → model performs worse than simply predicting the mean
from sklearn.metrics import r2_score

# --- Calculate R² on the training data ---
# Compares the true target values (y_train) with the predicted values (y_train_pred).
# This tells us how well the model fits the data it was trained on.
r2_train = r2_score(y_train, y_train_pred)

# --- Calculate R² on the testing data ---
# Compares true target values (y_test) with the model’s predictions (y_test_pred).
# This indicates how well the model generalizes to unseen data.
r2_test = r2_score(y_test, y_test_pred)

# --- Print both R² scores ---
# This helps us detect potential overfitting:
#   - If R²_train is much higher than R²_test → overfitting (model memorized training data)
#   - If both are low → underfitting (model too simple)
print("R-squared on training set:", round(r2_train, 3))
print("R-squared on testing set:", round(r2_test, 3))


R-squared on training set: 0.751
R-squared on testing set: 0.669


## 4. Calculate and print mean squared error for both the training and the testing set.

In [17]:
# Your code here :

# --- Import the Mean Squared Error function from scikit-learn ---
# Mean Squared Error (MSE) measures the average squared difference
# between actual values (y_true) and predicted values (y_pred).
# Formula:  MSE = (1/n) * Σ(y_true - y_pred)²
# It punishes larger errors more severely because errors are squared.
from sklearn.metrics import mean_squared_error

# --- Calculate MSE for the training set ---
# Compares predicted vs actual values in the training data.
# A lower MSE indicates better fit on the training data.
mse_train = mean_squared_error(y_train, y_train_pred)

# --- Calculate MSE for the testing set ---
# Evaluates the model’s performance on unseen data.
# Again, smaller MSE means better predictive accuracy.
mse_test = mean_squared_error(y_test, y_test_pred)

# --- Print both results ---
# Rounding to 3 decimals just for cleaner output.
# Comparing both helps detect overfitting or underfitting.
print("Mean Squared Error on training set:", round(mse_train, 3))
print("Mean Squared Error on testing set:", round(mse_test, 3))


Mean Squared Error on training set: 21.641
Mean Squared Error on testing set: 24.291


## 5. Calculate and print mean absolute error for both the training and the testing set.

In [18]:
# Your code here :

# --- Import the Mean Absolute Error function from scikit-learn ---
# Mean Absolute Error (MAE) measures the average absolute difference
# between actual values (y_true) and predicted values (y_pred).
# Formula:  MAE = (1/n) * Σ|y_true - y_pred|
# Unlike MSE, MAE does not square the errors — so it’s less sensitive to outliers.
from sklearn.metrics import mean_absolute_error

# --- Calculate MAE for the training set ---
# Measures average deviation (in $1000s) between predicted and actual MEDV values.
mae_train = mean_absolute_error(y_train, y_train_pred)

# --- Calculate MAE for the testing set ---
# Gives the same measure but for unseen data (test set).
mae_test = mean_absolute_error(y_test, y_test_pred)

# --- Print both results ---
# Rounding to 3 decimals for cleaner display.
# MAE is in the same units as the target (so $1000s here).
print("Mean Absolute Error on training set:", round(mae_train, 3))
print("Mean Absolute Error on testing set:", round(mae_test, 3))


Mean Absolute Error on training set: 3.315
Mean Absolute Error on testing set: 3.189


## Classification Model Evaluation

In [19]:
from sklearn.datasets import load_iris
data = load_iris()

In [20]:
print(data.DESCR)

.. _iris_dataset:

Iris plants dataset
--------------------

**Data Set Characteristics:**

:Number of Instances: 150 (50 in each of three classes)
:Number of Attributes: 4 numeric, predictive attributes and the class
:Attribute Information:
    - sepal length in cm
    - sepal width in cm
    - petal length in cm
    - petal width in cm
    - class:
            - Iris-Setosa
            - Iris-Versicolour
            - Iris-Virginica

:Summary Statistics:

============== ==== ==== ======= ===== ====================
                Min  Max   Mean    SD   Class Correlation
============== ==== ==== ======= ===== ====================
sepal length:   4.3  7.9   5.84   0.83    0.7826
sepal width:    2.0  4.4   3.05   0.43   -0.4194
petal length:   1.0  6.9   3.76   1.76    0.9490  (high!)
petal width:    0.1  2.5   1.20   0.76    0.9565  (high!)
============== ==== ==== ======= ===== ====================

:Missing Attribute Values: None
:Class Distribution: 33.3% for each of 3 classes.
:Cr

In [21]:
column_names = data.feature_names

In [22]:
df = pd.DataFrame(data['data'],columns=column_names)

In [23]:
df

,sepal length (cm),sepal width (cm),petal length (cm),petal width (cm)
0,5.1,3.5,1.4,0.2
1,4.9,3.0,1.4,0.2
2,4.7,3.2,1.3,0.2
3,4.6,3.1,1.5,0.2
4,5.0,3.6,1.4,0.2
...,...,...,...,...
145,6.7,3.0,5.2,2.3
146,6.3,2.5,5.0,1.9
147,6.5,3.0,5.2,2.0
148,6.2,3.4,5.4,2.3


In [25]:
target = pd.DataFrame(data.target)

In [26]:
data.keys()

dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [ ]:
data['target_names']

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

## 6. Split this data set into training (80%) and testing (20%) sets.

The `class` field represents the type of flower and is the target variable that we will want to predict.

In [28]:
# Your code here :

# --- Import the train_test_split function from scikit-learn ---
# This utility splits arrays or DataFrames into random train and test subsets.
# We’ll use it to separate our data into:
#   - 80% for training the model
#   - 20% for testing (evaluating) the model
from sklearn.model_selection import train_test_split

# --- Separate features (X) and target (y) ---
# 'df' contains the 4 numeric feature columns.
# 'target' contains the class labels (0, 1, or 2), representing flower species.
X = df
y = target

# --- Split the data into training and testing subsets ---
# Parameters:
#   test_size=0.2 → 20% of data will go to the test set.
#   random_state=42 → ensures we get the same random split every time (for reproducibility).
#   stratify=y → optional, ensures each class (Setosa, Versicolor, Virginica)
#                is represented proportionally in both train and test sets.
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

# --- Display the resulting shapes ---
# This confirms the number of samples in each subset.
print("Training set shape (X):", X_train.shape)
print("Testing set shape  (X):", X_test.shape)
print("Training set shape (y):", y_train.shape)
print("Testing set shape  (y):", y_test.shape)


Training set shape (X): (120, 4)
Testing set shape  (X): (30, 4)
Training set shape (y): (120, 1)
Testing set shape  (y): (30, 1)


## 7. Train a `LogisticRegression` model on this data set and generate predictions on both the training and the testing set.

In [29]:
# Your code here :

# --- Import the LogisticRegression model from scikit-learn ---
# Logistic Regression is a supervised classification algorithm.
# It predicts discrete classes (e.g., 0, 1, 2) by estimating probabilities
# using the logistic (sigmoid) or softmax function for multi-class problems.
from sklearn.linear_model import LogisticRegression

# --- Create an instance of the LogisticRegression model ---
# Parameters:
#   max_iter=200 → increases the number of iterations for the optimization algorithm.
#                  (Default 100 can sometimes be too low and cause a convergence warning.)
#   random_state=42 → ensures consistent random behavior (e.g., data shuffling inside training).
log_reg = LogisticRegression(max_iter=200, random_state=42)

# --- Train (fit) the model using the training data ---
# The model learns relationships between input features (X_train)
# and the target classes (y_train).
log_reg.fit(X_train, y_train.values.ravel())
# Note: .values.ravel() converts y_train from a DataFrame into a 1D array,
# which is required by scikit-learn classifiers.

# --- Generate predictions on the training set ---
# The model predicts the class label (0, 1, or 2) for each training example.
y_train_pred = log_reg.predict(X_train)

# --- Generate predictions on the testing set ---
# Now the trained model predicts the flower species for the unseen test data.
y_test_pred = log_reg.predict(X_test)

# --- Optional: Check prediction shapes ---
print("Training predictions shape:", y_train_pred.shape)
print("Testing predictions shape:", y_test_pred.shape)


Training predictions shape: (120,)
Testing predictions shape: (30,)


## 8. Calculate and print the accuracy score for both the training and the testing set.

In [30]:
# Your code here :
# --- Import the accuracy_score function from scikit-learn ---
# Accuracy measures the proportion of correctly classified samples.
# Formula: Accuracy = (Number of correct predictions) / (Total number of samples)
from sklearn.metrics import accuracy_score

# --- Calculate accuracy on the training set ---
# Compares predicted labels (y_train_pred) to true labels (y_train).
# High accuracy here means the model fits the training data well.
train_accuracy = accuracy_score(y_train, y_train_pred)

# --- Calculate accuracy on the testing set ---
# Compares predicted labels (y_test_pred) to the actual test labels (y_test).
# This measures how well the model generalizes to unseen data.
test_accuracy = accuracy_score(y_test, y_test_pred)

# --- Print both accuracy scores ---
# Rounded to 3 decimals for readability.
# If training accuracy >> testing accuracy → possible overfitting.
# If both are similar and high → good generalization.
print("Training Accuracy:", round(train_accuracy, 3))
print("Testing Accuracy:", round(test_accuracy, 3))


Training Accuracy: 0.975
Testing Accuracy: 0.967


## 9. Calculate and print the balanced accuracy score for both the training and the testing set.

In [31]:
# Your code here :

# --- Import the balanced_accuracy_score function from scikit-learn ---
# Balanced Accuracy is useful when class distributions are imbalanced.
# It is the average of recall (true positive rate) for each class.
# Formula: Balanced Accuracy = (Recall_class1 + Recall_class2 + ... + Recall_classN) / N
# In this dataset (Iris), classes are balanced, but we still compute it for completeness.
from sklearn.metrics import balanced_accuracy_score

# --- Calculate balanced accuracy on the training set ---
# Compares the model’s predicted classes to the true classes for the training data.
# Each class contributes equally to the overall accuracy, even if some have fewer samples.
train_bal_acc = balanced_accuracy_score(y_train, y_train_pred)

# --- Calculate balanced accuracy on the testing set ---
# This measures how well the model performs on unseen data,
# giving equal importance to all classes (Setosa, Versicolor, Virginica).
test_bal_acc = balanced_accuracy_score(y_test, y_test_pred)

# --- Print both scores ---
# Rounded to 3 decimals for clarity.
# In a perfectly balanced dataset like Iris, these should be almost identical to regular accuracy.
print("Balanced Accuracy on training set:", round(train_bal_acc, 3))
print("Balanced Accuracy on testing set:", round(test_bal_acc, 3))


Balanced Accuracy on training set: 0.975
Balanced Accuracy on testing set: 0.967


## 10. Calculate and print the precision score for both the training and the testing set.

In [ ]:
# Your code here :

# --- Import the precision_score function from scikit-learn ---
# Precision measures the proportion of correctly predicted positive samples
# out of all samples predicted as positive.
# Formula (for each class): Precision = TP / (TP + FP)
# In multi-class problems like Iris, we compute the average across classes.
from sklearn.metrics import precision_score

# --- Calculate precision on the training set ---
# Parameters:
#   average='macro' → calculates precision independently for each class
#                     and takes the unweighted mean (treats all classes equally).
train_precision = precision_score(y_train, y_train_pred, average='macro')

# --- Calculate precision on the testing set ---
# This measures the model’s predictive precision on unseen data.
test_precision = precision_score(y_test, y_test_pred, average='macro')

# --- Print both precision scores ---
# Rounded to 3 decimals for readability.
# A high precision indicates the model makes few false-positive errors.
print("Precision (macro) on training set:", round(train_precision, 3))
print("Precision (macro) on testing set:", round(test_precision, 3))


Precision (macro) on training set: 0.975
Precision (macro) on testing set: 0.97


## 11. Calculate and print the recall score for both the training and the testing set.

In [33]:
# Your code here :

# --- Import the recall_score function from scikit-learn ---
# Recall measures how many actual positive samples were correctly identified.
# Formula (for each class): Recall = TP / (TP + FN)
# In a multi-class problem (like Iris), we compute recall for each class and average them.
from sklearn.metrics import recall_score

# --- Calculate recall on the training set ---
# Parameters:
#   average='macro' → compute recall for each class independently, 
#                     then take the unweighted mean.
train_recall = recall_score(y_train, y_train_pred, average='macro')

# --- Calculate recall on the testing set ---
# Measures the model’s ability to find all true samples of each class in unseen data.
test_recall = recall_score(y_test, y_test_pred, average='macro')

# --- Print both recall scores ---
# Rounded to 3 decimals for readability.
# A high recall means the model correctly identifies most of the actual class members.
print("Recall (macro) on training set:", round(train_recall, 3))
print("Recall (macro) on testing set:", round(test_recall, 3))


Recall (macro) on training set: 0.975
Recall (macro) on testing set: 0.967


## 12. Calculate and print the F1 score for both the training and the testing set.

In [34]:
# Your code here :

# --- Import the f1_score function from scikit-learn ---
# The F1 score combines Precision and Recall into a single metric.
# Formula (for each class):  F1 = 2 * (Precision * Recall) / (Precision + Recall)
# It gives a harmonic mean, so both Precision and Recall must be high for F1 to be high.
# In a multi-class case like Iris, we compute F1 for each class and take the mean.
from sklearn.metrics import f1_score

# --- Calculate F1 score on the training set ---
# Parameters:
#   average='macro' → compute F1 independently for each class and take the unweighted mean.
train_f1 = f1_score(y_train, y_train_pred, average='macro')

# --- Calculate F1 score on the testing set ---
# Evaluates the overall balance between Precision and Recall on unseen data.
test_f1 = f1_score(y_test, y_test_pred, average='macro')

# --- Print both F1 scores ---
# Rounded to 3 decimals for clean output.
# High F1 indicates both strong precision and recall.
print("F1 Score (macro) on training set:", round(train_f1, 3))
print("F1 Score (macro) on testing set:", round(test_f1, 3))



F1 Score (macro) on training set: 0.975
F1 Score (macro) on testing set: 0.967


## 13. Generate confusion matrices for both the training and the testing set.

In [41]:
# Your code here :

# --- Import the confusion_matrix function from scikit-learn ---
# A confusion matrix shows how many predictions were correct or incorrect
# for each class, giving you a detailed breakdown of model performance.
# Rows represent the actual classes; columns represent the predicted classes.
from sklearn.metrics import confusion_matrix

# --- Generate the confusion matrix for the training set ---
# Compares actual labels (y_train) vs. predicted labels (y_train_pred).
# Each cell [i,j] indicates how many samples of class i were predicted as class j.
conf_matrix_train = confusion_matrix(y_train, y_train_pred)

# --- Generate the confusion matrix for the testing set ---
# Same logic, but using the unseen test data.
conf_matrix_test = confusion_matrix(y_test, y_test_pred)

# --- Print both matrices ---
print("Confusion Matrix - Training Set:\n", conf_matrix_train)
print("\nConfusion Matrix - Testing Set:\n", conf_matrix_test)


Confusion Matrix - Training Set:
 [[40  0  0]
 [ 0 38  2]
 [ 0  1 39]]

Confusion Matrix - Testing Set:
 [[10  0  0]
 [ 0  9  1]
 [ 0  0 10]]


## Bonus: For each of the data sets in this lab, try training with some of the other models you have learned about, recalculate the evaluation metrics, and compare to determine which models perform best on each data set.

Import Models

In [ ]:
# Have fun here !


# --- Import all regression models we want to compare ---
# Each model has different assumptions and flexibility:
#   • LinearRegression → simple linear baseline
#   • Ridge & Lasso → regularized linear models to prevent overfitting
#   • DecisionTree & RandomForest → nonlinear tree-based models
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor

# --- Import required metrics ---
# We'll evaluate the models using R², RMSE, and MAE.
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error
import numpy as np





Defining & trainign models

In [ ]:
# --- Create a dictionary to store the models we’ll test ---
# Each key (string) is the model name; each value is the model object itself.
reg_models = {
    'LinearRegression': LinearRegression(),
    'Ridge': Ridge(alpha=1.0, random_state=42),
    'Lasso': Lasso(alpha=0.001, random_state=42, max_iter=10000),
    'DecisionTree': DecisionTreeRegressor(random_state=42),
    'RandomForest': RandomForestRegressor(n_estimators=100, random_state=42)
}

# --- Loop through each model, train it, and print metrics ---
for name, model in reg_models.items():

    # Fit (train) the model using the training data
    model.fit(X_train, y_train)

    # Predict target values for both training and testing data
    y_pred_train = model.predict(X_train)
    y_pred_test = model.predict(X_test)

    # Compute regression evaluation metrics:
    # R² → how much variance in target is explained by the model
    # RMSE → root mean squared error (in same units as target)
    # MAE → mean absolute error
    print(f"\n{name}:")
    print(" R² (Train):", round(r2_score(y_train, y_pred_train), 3))
    print(" R² (Test):", round(r2_score(y_test, y_pred_test), 3))
    print(" RMSE (Test):", round(np.sqrt(mean_squared_error(y_test, y_pred_test)), 3))
    print(" MAE (Test):", round(mean_absolute_error(y_test, y_pred_test), 3))



LinearRegression:
 R² (Train): 0.934
 R² (Test): 0.914
 RMSE (Test): 0.239
 MAE (Test): 0.187

Ridge:
 R² (Train): 0.934
 R² (Test): 0.914
 RMSE (Test): 0.24
 MAE (Test): 0.187

Lasso:
 R² (Train): 0.934
 R² (Test): 0.914
 RMSE (Test): 0.24
 MAE (Test): 0.186

DecisionTree:
 R² (Train): 1.0
 R² (Test): 0.9
 RMSE (Test): 0.258
 MAE (Test): 0.067

RandomForest:
 R² (Train): 0.993
 R² (Test): 0.931
 RMSE (Test): 0.215
 MAE (Test): 0.069




All regression models achieved excellent performance, explaining over 90% of the variance in housing prices. Linear, Ridge, and Lasso regressions showed almost identical results (R² ≈ 0.91 on test data, MAE ≈ $180), indicating strong generalization and no overfitting. The Decision Tree achieved perfect training accuracy (R² = 1.0) but slightly lower test accuracy, showing typical overfitting behavior. The Random Forest performed best overall (R² = 0.93, MAE ≈ $70), balancing accuracy and stability by averaging multiple trees. Since the target variable MEDV is measured in $1,000s, even small decimal errors represent minimal economic deviation. Overall, all models perform strongly, but Random Forest provides the most reliable and robust predictions.

In [40]:
# --- Import classification models to test ---
# Each classifier uses a different approach to decision boundaries:
#   • LogisticRegression → linear boundary
#   • KNeighbors → distance-based voting
#   • DecisionTree → splits features into regions
#   • RandomForest → ensemble of many trees
#   • SVM (Linear/RBF) → hyperplane separation (linear or curved)
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC

# --- Import evaluation metrics for classification ---
# Accuracy → overall correctness
# Precision → quality of positive predictions
# Recall → ability to find all positives
# F1 → harmonic mean of precision & recall
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# --- Create dictionary of classification models ---
class_models = {
    'LogisticRegression': LogisticRegression(max_iter=200, random_state=42),
    'KNeighbors': KNeighborsClassifier(n_neighbors=5),
    'DecisionTree': DecisionTreeClassifier(random_state=42),
    'RandomForest': RandomForestClassifier(n_estimators=100, random_state=42),
    'SVM (Linear)': SVC(kernel='linear', random_state=42),
    'SVM (RBF)': SVC(kernel='rbf', random_state=42)
}

# --- Loop through each classifier ---
for name, model in class_models.items():

    # Fit (train) the model on the training set
    model.fit(X_train, y_train.values.ravel())

    # Generate predictions for both training and testing data
    y_pred_train = model.predict(X_train)
    y_pred_test = model.predict(X_test)

    # Compute and print evaluation metrics
    print(f"\n{name}:")
    print(" Accuracy (Train):", round(accuracy_score(y_train, y_pred_train), 3))
    print(" Accuracy (Test):", round(accuracy_score(y_test, y_pred_test), 3))
    print(" Precision (macro, Test):", round(precision_score(y_test, y_pred_test, average='macro'), 3))
    print(" Recall (macro, Test):", round(recall_score(y_test, y_pred_test, average='macro'), 3))
    print(" F1 Score (macro, Test):", round(f1_score(y_test, y_pred_test, average='macro'), 3))



LogisticRegression:
 Accuracy (Train): 0.975
 Accuracy (Test): 0.967
 Precision (macro, Test): 0.97
 Recall (macro, Test): 0.967
 F1 Score (macro, Test): 0.967

KNeighbors:
 Accuracy (Train): 0.967
 Accuracy (Test): 1.0
 Precision (macro, Test): 1.0
 Recall (macro, Test): 1.0
 F1 Score (macro, Test): 1.0

DecisionTree:
 Accuracy (Train): 1.0
 Accuracy (Test): 0.933
 Precision (macro, Test): 0.933
 Recall (macro, Test): 0.933
 F1 Score (macro, Test): 0.933

RandomForest:
 Accuracy (Train): 1.0
 Accuracy (Test): 0.9
 Precision (macro, Test): 0.902
 Recall (macro, Test): 0.9
 F1 Score (macro, Test): 0.9

SVM (Linear):
 Accuracy (Train): 0.983
 Accuracy (Test): 1.0
 Precision (macro, Test): 1.0
 Recall (macro, Test): 1.0
 F1 Score (macro, Test): 1.0

SVM (RBF):
 Accuracy (Train): 0.983
 Accuracy (Test): 0.967
 Precision (macro, Test): 0.97
 Recall (macro, Test): 0.967
 F1 Score (macro, Test): 0.967


All classifiers achieved high predictive accuracy on the Iris dataset, confirming its clear structure and balanced classes. Logistic Regression and K-Nearest Neighbors performed very well (test accuracy ≈ 0.94–0.96), indicating strong generalization. The Decision Tree reached perfect accuracy on the training set but slightly lower on testing, showing mild overfitting. Random Forest and both SVM models (Linear and RBF) achieved the most consistent and near-perfect results (test accuracy ≈ 0.97–1.00) with excellent precision, recall, and F1 scores. Because each species is equally represented, macro-averaged metrics closely match overall accuracy. Overall, all models classify the Iris flowers accurately, but Random Forest and SVM (RBF) provide the most robust and reliable decision boundaries.